### PR TITLE
#120 Upgrading Mongo java driver to 4.1.1 from 3.12

### PR DESCRIPTION
**Pull request by buchi-busireddy**
https://github.com/hypertrace/hypertrace/issues/120

### CODE REVIEW

File: document-store/build.gradle.kts [hunk 1]
```diff
@@ -16,7 +16,7 @@ plugins {
 
 dependencies {
   api("com.typesafe:config:1.3.2")
-  implementation("org.mongodb:mongo-java-driver:3.12.0")
+  implementation("org.mongodb:mongodb-driver-sync:4.1.1")
```
Comment (tim-mwangi) [inline document-store/build.gradle.kts:19]: Was this async or have we been using sync mongo?

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java [hunk 2]
```diff
@@ -312,15 +321,13 @@ public long total(Query query) {
...
-    return collection.count(ref);
+    return collection.countDocuments(new BasicDBObject(map));
   }
 
   @Override
   public boolean bulkUpsert(Map<Key, Document> documents) {
     try {
-      BulkWriteOperation bulkWriteOperation = collection.initializeUnorderedBulkOperation();
-
+      List<UpdateOneModel<BasicDBObject>> bulkCollection = new ArrayList<>();
```
Comment (tim-mwangi) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java:330]: Checked the codecov for this. There's no coverage going through bulkUpsert?

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java [hunk 3]
```diff
@@ -46,33 +50,38 @@
   private static final String CREATED_TIME = "createdTime";
   private static final ObjectMapper MAPPER = new ObjectMapper();
 
-  private final DBCollection collection;
+  private final com.mongodb.client.MongoCollection<BasicDBObject> collection;
 
-  MongoCollection(DBCollection collection) {
+  MongoCollection(com.mongodb.client.MongoCollection<BasicDBObject> collection) {
```
Comment (tim-mwangi) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java:55]: Something wrong with having this among the imports?

Comment (tim-mwangi) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java:55]: Oh probably coz this class is also called MongoCollection. Got it.

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoDatastore.java [hunk 4]
```diff
@@ -1,79 +1,92 @@
...
-        String hostName = config.getString("host");
-        int port = config.getInt("port");
-        client = new MongoClient(hostName, port);
-      }
-    } catch (IllegalArgumentException e) {
-        throw new IllegalArgumentException(
-            String.format("Unable to instantiate MongoClient with config:%s",config), e);
+    if (config.hasPath("url")) {
+      ConnectionString connString = new ConnectionString(config.getString("url"));
+      MongoClientSettings settings = MongoClientSettings.builder()
```
Comment (tim-mwangi) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoDatastore.java:32]: This seems duplicated

File: document-store/build.gradle.kts (hunk 1 above)
Comment (buchi-busireddy) [reply document-store/build.gradle.kts:19]: We have always been using sync version, which is now their legacy driver. Starting 3.x, they separated the sync and reactive java drivers.

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java (hunk 3 above)
Comment (buchi-busireddy) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java:55]: It conflicts with the class name :)

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoDatastore.java (hunk 4 above)
Comment (buchi-busireddy) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoDatastore.java]: Cleaned up.

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java (hunk 2 above)
Comment (buchi-busireddy) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java:330]: Added now.

File: document-store/build.gradle.kts (hunk 1 above)
Comment (tim-mwangi) [reply document-store/build.gradle.kts:19]: We should probably switch to reactive at some point

**Review by tim-mwangi — APPROVED** [review]

File: document-store/build.gradle.kts (hunk 1 above)
Comment (inespo) [reply document-store/build.gradle.kts:19]: Interesting proposal to have async for data persistence. I would like to understand how we're going to handle data inconsistency, error handling, and our SLA (is eventual consistency acceptable) with having async data persistence. But, I think it's outside the scope of this review.

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoDatastore.java [hunk 5]
```diff
@@ -1,79 +1,89 @@
...
-        client.getDB(DEFAULT_DB_NAME).createCollection(collectionName, dbObject);
-
-    return collection != null;
+    try {
+      database.createCollection(collectionName);
+    } catch (MongoCommandException e) {
+      if (!"NamespaceExists".equals(e.getErrorCodeName())) {
+        LOGGER.error("Could not create collection: {}", collectionName, e);
+        return false;
+      }
```
Comment (inespo) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoDatastore.java:66]: I think it's still worth Logging the error for other failures too for debugging purpose and not limited to just NamespaceExists

**Review by inespo — APPROVED** [review]

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoDatastore.java (hunk 5 above)
Comment (buchi-busireddy) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoDatastore.java:66]: I'm logging it for other cases itself @inespo

Comment (inespo) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoDatastore.java:66]: ah yes, missed the "!"

**Review by inespo — APPROVED** [review]

**Comment by buchi-busireddy** [comment]
closing in favor of https://github.com/hypertrace/document-store/pull/12

File: document-store/build.gradle.kts (hunk 1 above)
Comment (tim-mwangi) [reply document-store/build.gradle.kts:19]: Async is about data transfer not persistence. The client would be async.

**Pull request closed without merging**